### DISCUSSION TITLE
How to add time field

### DISCUSSION THREAD

**Discussion by uglyliu** [vectordotdev/vector#17250] [Q&A] [answered]
# Help me please
# Architecture
- Filebeat-->Kafka-->Vector-->ES
# Version
- vector 0.28.1 (x86_64-unknown-linux-gnu ff15924 2023-03-06)
# Content
- How to customize a time field currentdate with a value like 2023.04.30 to replace %Y.%m.%d, and then reference it in the sink
# Config
[sinks.elasticsearch]
 type = "elasticsearch" # required
 inputs = ["parse_logs"] # required
 api_version = "auto"
 compression = "none" # optional, default
 endpoints = ["http://127.0.0.1:9200"] # required
 mode = "bulk"
 bulk.action = "index"
 bulk.index = "vector-java-test-%Y.%m.%d" 

 #healthcheck.enabled = true # optional, default

**Comment by spencergilbert** [comment]
What is the behavior of the configuration you provided? The documentation for [templates](https://vector.dev/docs/reference/configuration/template-syntax/) and the [elasticsearch](https://vector.dev/docs/reference/configuration/sinks/elasticsearch/#bulk.index) sink suggest it should work.

**Reply by uglyliu** [reply]
> What is the behavior of the configuration you provided? The documentation for [templates](https://vector.dev/docs/reference/configuration/template-syntax/) and the [elasticsearch](https://vector.dev/docs/reference/configuration/sinks/elasticsearch/#bulk.index) sink suggest it should work.

I want to replace %Y.%m.%d with a custom time field such as 2023.04.30, and then I reference this field in the sink configuration，Then I don't know how to deal with it, I'm not very familiar with vector，Can you help me please

**Reply by spencergilbert** [reply]
Have you run the configuration you provided, the configuration you provided should add the current time, formatted, to the index.

**Reply by uglyliu** [reply]
> Have you run the configuration you provided, the configuration you provided should add the current time, formatted, to the index.


Yes, that can indeed create the date of the day, but when elasticsearch creates the index, it will create the new index at 8 o'clock in the morning the next day, so I want to customize a time field without the built-in format, and I do the same in logstash made,my logstash  config as follows:

date {
            match => [ "timestamp", "YYYY-MM-dd HH:mm:ss", "ISO8601" ]
            ## 2022-07-22 11:00:02
        }
ruby {
         code => "event.set('currentdate', event.get('@timestamp').time.localtime + 8*60*60)"
    }
mutate {
         convert => ["currentdate", "string"]
         gsub => ["currentdate", "T([\S\s]*?)Z", ""]
         gsub => ["currentdate", "-", "."]
    }
output {
  elasticsearch {
    hosts => [*******]
    index => "XXXX-%{currentdate}"
}
}

**Reply by spencergilbert** [reply]
I'm not sure how Elasticsearch creating indexes comes into play here, if it's set to auto-create indices Vector's templating of the time is based on it's timezone.

You could also convert your example Logstash config into a `remap` component.

**Reply by uglyliu** [reply]
> I'm not sure how Elasticsearch creating indexes comes into play here, if it's set to auto-create indices Vector's templating of the time is based on it's timezone.
> 
> You could also convert your example Logstash config into a `remap` component.

Maybe the config is effect, I will look at the es index tomorrow,the timestamp is in the form of `2023-05-03 13:49:54.004` 

```
[transforms.parse_logs]
type = "remap"
inputs = ["dummy_logs"]
source        = """
     .currentdate = split(replace(.timestamp,"-",".")," ",limit: 2)[0]
"""

[sinks.elasticsearch]
 bulk.index = "vector-accessjava-{{ .currentdate }}"

```

**Reply by spencergilbert** [reply]
You should use [to_timestamp()](https://vector.dev/docs/reference/vrl/functions/#to_timestamp) and either [parse_timestamp()](https://vector.dev/docs/reference/vrl/functions/#parse_timestamp) or [format_timestamp()](https://vector.dev/docs/reference/vrl/functions/#format_timestamp) rather than trying to manipulate it as a string.

It should be less error prone.

**Reply by uglyliu** [reply]
> You should use [to_timestamp()](https://vector.dev/docs/reference/vrl/functions/#to_timestamp) and either [parse_timestamp()](https://vector.dev/docs/reference/vrl/functions/#parse_timestamp) or [format_timestamp()](https://vector.dev/docs/reference/vrl/functions/#format_timestamp) rather than trying to manipulate it as a string.
> 
> It should be less error prone.

Thank you for your suggestion. I will try format_timestamp and parse_timestamp today. My method is currently working. Customizing the time suffix of the es index can prevent the es from creating the index 8 hours later.
![image](https://user-images.githubusercontent.com/22983421/236066476-ab091de6-a16a-4afa-85c4-14bc3e2f7b64.png)

**Reply by uglyliu** [reply, marked as answer]
> You should use [to_timestamp()](https://vector.dev/docs/reference/vrl/functions/#to_timestamp) and either [parse_timestamp()](https://vector.dev/docs/reference/vrl/functions/#parse_timestamp) or [format_timestamp()](https://vector.dev/docs/reference/vrl/functions/#format_timestamp) rather than trying to manipulate it as a string.
> 
> It should be less error prone.

use to_timestamp and format_timestamp is done, thank u very much

```
.timestamp = now()
.unix_time,error = to_int(format_timestamp!(.timestamp, format: "%s"))
.utc8_time = .unix_time
.currentdate = format_timestamp!(to_timestamp!(.utc8_time), format: "%Y.%m.%d")
```
